### PR TITLE
Switch to maven publish plugin for publishing

### PR DESCRIPTION
**Pull request by ZacSweers**
This migrates the project to use the maven publish plugin and is functionally equivalent to the artifacts published with the previous plugin infrastructure.

Also opportunistically updates gradle, develocity, and kotlin along the way

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -54,10 +55,6 @@ publishing {
       artifactId = 'sort-gradle-dependencies-dist'
       artifact shadowDistZip
     }
-    app(MavenPublication) {
-      artifactId = 'sort-gradle-dependencies-app'
```
Comment (ZacSweers) [inline app/build.gradle:58]: This one is created automatically by the plugin

File: build-logic/src/main/kotlin/com/squareup/convention/LibConvention.kt [hunk 2]
```diff
@@ -16,11 +16,5 @@ class LibConvention : Plugin<Project> {
 
   override fun apply(target: Project): Unit = target.run {
     pluginManager.apply(BaseConvention::class.java)
-
-    extensions.configure(JavaPluginExtension::class.java) { j ->
-      // We specifically don't want either of these artifacts on the app project
```
Comment (ZacSweers) [inline build-logic/src/main/kotlin/com/squareup/convention/LibConvention.kt:21]: Handled automatically by the plugin when dokka is present

**Comment by autonomousapps** [comment]
This is almost ready to merge.

- [ ] Need to validate the GHA/yaml changes.

File: .github/workflows/publish.yml [hunk 3]
```diff
@@ -36,8 +36,11 @@ jobs:
...
         with:
-          arguments: publishToMavenCentral -s
+          arguments: publishToMavenCentral --no-configuration-cache
         env:
           ORG_GRADLE_PROJECT_mavenCentralUsername: ${{ secrets.SONATYPE_NEXUS_USERNAME }}
           ORG_GRADLE_PROJECT_mavenCentralPassword: ${{ secrets.SONATYPE_NEXUS_PASSWORD }}
           ORG_GRADLE_PROJECT_signingInMemoryKey: ${{ secrets.ARTIFACT_SIGNING_PRIVATE_KEY }}
+          # TODO from the square side
+          ORG_GRADLE_PROJECT_signingInMemoryKeyId: ${{ secrets.TODO }}
+          ORG_GRADLE_PROJECT_signingInMemoryKeyPassword: ${{ secrets.TODO }}
```
Comment (autonomousapps) [inline .github/workflows/publish.yml:46]: Validate.

**Review by autonomousapps — APPROVED** [review]

**Pull request merged**